### PR TITLE
Simplify Spin implementation

### LINKED ISSUE #14100: Spin 'delay' prop
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[https://codepen.io/xusiyuan841028-1471240073/pen/YdaXry?&editable=true&editors=001](https://codepen.io/xusiyuan841028-1471240073/pen/YdaXry?&editable=true&editors=001)

### Steps to reproduce
If set up 'delay' prop, Spin won't display!

### What is expected?
Spin will display after 200ms.

### What is actually happening?
Spin is always hidden.

| Environment | Info |
|---|---|
| antd | 3.11.6 |
| React | 16.7.0 |
| System | OSX |
| Browser | Chrome |

---
This [commit](https://github.com/ant-design/ant-design/commit/2c7112be7bf32c6e8362334b86b0799cc3a4a6c4#r3184044) brought out this issue.

### PR DESCRIPTION
**Pull request by afc163**
First of all, thank you for your contribution! 😄

Please makes sure that these form are filled before submitting your pull request, thank you!

### This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / document update
- [ ] Branch merge
- [ ] Other (about what?)

### What's the background?
  
close #14100
  
### What's the effect? (Optional if not new feature)

changelog: 🐛  Fixed spin with delay never renders. #14100

### Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog provided

### Additional Plan? (Optional if not new feature)

> If this PR related with other PR or following info. You can type here.

### CODE REVIEW

**Comment by zombieJ** [comment]
汗，原来你也在搞……把我这个 test 合进来吧：https://github.com/ant-design/ant-design/blob/spinDelay/components/spin/__tests__/index.test.js#L51

File: components/spin/index.tsx [hunk 1]
```diff
@@ -78,62 +79,43 @@ class Spin extends React.Component<SpinProps, SpinState> {
...
 
     const { spinning, delay } = props;
+    const shouldBeDelayed = shouldDelay(spinning, delay);
     this.state = {
-      spinning: spinning && !shouldDelay(spinning, delay),
+      spinning: spinning && !shouldBeDelayed,
     };
+
+    if (shouldBeDelayed) {
+      this.updateSpinning = debounce(this.updateSpinning, delay);
```
Comment (zombieJ) [inline components/spin/index.tsx]: 这么写官方例子就挂了……还是需要在 update 的时候判断做 delay

**Comment by afc163** [comment]
Fixed

**Comment by picodoth** [comment]
没看到，把另一个合了, rebase 一下

File: components/spin/index.tsx (hunk 1 above)
Comment (picodoth) [reply components/spin/index.tsx]: @zombieJ 这个还有问题么？

**Review by picodoth — APPROVED** [review]

**Comment by picodoth** [comment]
@zombieJ 再看下么

File: components/spin/index.tsx (hunk 1 above)
Comment (zombieJ) [reply components/spin/index.tsx]: 如果 `delay` 是后面变更的，这个 `delay` 就失效了。
另外这个写法使得 `delay` 这个[例子](https://ant.design/components/spin-cn/#components-spin-demo-delayAndDebounce) 的停止 spinning 也有延迟了：
kapture 2019-01-06 at 11 40 24

Comment (afc163) [reply components/spin/index.tsx]: 停止 spinning 本来的设计就有延迟。

Comment (afc163) [reply components/spin/index.tsx]: delay 变更毫无实际意义，所以做的时候就不考虑。

Comment (ztplz) [reply components/spin/index.tsx]: 这个延迟可以交给用户啊

Comment (zombieJ) [reply components/spin/index.tsx]: > delay 变更毫无实际意义，所以做的时候就不考虑。

原本支持的话，改动应该还是能支持才行。

File: components/spin/index.tsx [hunk 2]
```diff
@@ -78,66 +79,47 @@ class Spin extends React.Component<SpinProps, SpinState> {
...
-      if (shouldDelay(spinning, delay)) {
-        if (this.delayTimeout) {
-          clearTimeout(this.delayTimeout);
-        }
-        this.delayTimeout = window.setTimeout(this.delayUpdateSpinning, delay);
-      } else {
-        this.setState({ spinning });
-      }
+    if (delay) {
+      this.updateSpinning = debounce(this.updateSpinning, delay);
```
Comment (zombieJ) [inline components/spin/index.tsx]: 抽一个 updateSpinning 的再 debounce，否则每次 update 都嵌一层……

Comment (afc163) [reply components/spin/index.tsx]: 👍

**Review by zombieJ — APPROVED** [review]

**Pull request merged**